### PR TITLE
#1573 - Integration of the stats monkey into the framework

### PR DESCRIPTION
**Pull request by kvanhijf**

### CODE REVIEW

File: webapps/frontend/app/viewmodels/site/admin/support.js [hunk 1]
```diff
@@ -76,55 +60,117 @@ define([
...
-                    .fail(function() {
+                    .fail(function(error) {
                         generic.alertError(
-                            $.t('ovs:support.failed'),
-                            $.t('ovs:support.failedmsg')
+                            $.t('ovs:support.settings.failed'),
+                            $.t('ovs:support.settings.failed_msg', {error: generic.extractErrorMessage(error)})
                         );
+                        // In case of failure, remove the selected setting again
+                        if (self.origStatsMonkeyConfig.isInitialized() === false) {
```
Comment (JeffreyDevloo) [inline webapps/frontend/app/viewmodels/site/admin/support.js]: elf.selectedSupportSettings should be self.oldSupportSettings(); when there is an error. This way all settings get reverted

File: ovs/extensions/support/agent.py [hunk 2]
```diff
@@ -41,53 +44,46 @@ def __init__(self):
...
-            data['errors'].append(str(ex))
+            for service_info in sorted(self._service_manager.list_services(client=self._client, add_status_info=True)):
+                if not service_info.startswith('ovs-'):
+                    continue
+                service_name = service_info.split()[0].strip()
+                services[service_name] = ' '.join(service_info.split()[1:])
+        except Exception as ex:
+            errors.append(str(ex))
+
+        data = {'cid': self._cluster_id,
```
Comment (JeffreyDevloo) [inline ovs/extensions/support/agent.py:81]: I'd opt for cluster_id and node_id as I had to ask what these meant

File: ovs/extensions/support/agent.py [hunk 3]
```diff
@@ -102,36 +98,35 @@ def _update_config(key, value):
...
-    @staticmethod
-    def _process_task(task, metadata, servicemanager):
+    def _process_task(self, task_code, metadata):
         """
         Processes a task
         """
         try:
-            SupportAgent._logger.debug('Processing: {0}'.format(task))
+            SupportAgent._logger.debug('Processing: {0}'.format(task_code))
             cid = Configuration.get('/ovs/framework/cluster_id').replace(r"'", r"'\''")
```
Comment (JeffreyDevloo) [inline ovs/extensions/support/agent.py]: Use self._cluster_id here

File: ovs/extensions/support/agent.py [hunk 4]
```diff
@@ -102,36 +98,35 @@ def _update_config(key, value):
...
-            nid = System.get_my_machine_id().replace(r"'", r"'\''")
+            nid = self._storagerouter.machine_id.replace(r"'", r"'\''")
 
-            if task == 'OPEN_TUNNEL':
-                if servicemanager == 'upstart':
+            if task_code == 'OPEN_TUNNEL':
+                if self._service_type == 'upstart':
                     check_output('service openvpn stop', shell=True)
```
Comment (JeffreyDevloo) [inline ovs/extensions/support/agent.py:112]: Screw you, other vpn user!

File: ovs/extensions/support/agent.py [hunk 5]
```diff
@@ -102,36 +98,35 @@ def _update_config(key, value):
...
                 else:
                     check_output("systemctl stop 'openvpn@ovs_{0}-{1}' || true".format(cid, nid), shell=True)
```
Comment (JeffreyDevloo) [inline ovs/extensions/support/agent.py]: Use the service manager here

File: ovs/extensions/support/agent.py [hunk 6]
```diff
@@ -143,7 +138,7 @@ def _process_task(task, metadata, servicemanager):
             else:
                 raise RuntimeError('Unknown task')
         except Exception, ex:
-            SupportAgent._logger.exception('Unexpected error while processing task {0} (data: {1}): {2}'.format(task, json.dumps(metadata), ex))
+            SupportAgent._logger.exception('Unexpected error while processing task {0} (data: {1}): {2}'.format(task_code, json.dumps(metadata), ex))
```
Comment (JeffreyDevloo) [inline ovs/extensions/support/agent.py]: Ex is not needed here

File: ovs/lib/storagerouter.py [hunk 7]
```diff
@@ -1429,38 +1438,139 @@ def get_proxy_config(vpool_guid, storagerouter_guid):
...
+            {'stats_monkey': True,  # Enable/disable the stats monkey scheduled task
+             'support_agent': True,  # Responsible for enabling the ovs-support-agent service, which collects heart beat data
+             'remote_access': False,  # Cannot be True when support agent is False. Is responsible for opening an OpenVPN tunnel to allow for remote access
+             'stats_monkey_config': {}}  # Dict with information on how to configure the stats monkey (Only required when enabling the stats monkey
+        :type support_info: dict
+        :return: None
+        :rtype: NoneType
         """
-        clients = []
+        Toolbox.verify_required_params(verify_keys=True,
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py]: The API shouldn't punish those that send to much

File: ovs/lib/storagerouter.py [hunk 8]
```diff
@@ -1429,38 +1438,139 @@ def get_proxy_config(vpool_guid, storagerouter_guid):
...
-                    service_manager.remove_service(StorageRouterController.SUPPORT_AGENT, client=root_client)
-        return True
+        if remote_access_change is True:
+            Configuration.set(key=remote_access_key, value=remote_access_new)
+            cid = Configuration.get('/ovs/framework/cluster_id').replace(r"'", r"'\''")
+            for storagerouter, root_client in root_clients.iteritems():
+                if remote_access_new is False:
+                    StorageRouterController._logger.info('Un-configuring remote access on StorageRouter {0}'.format(root_client.ip))
+                    nid = storagerouter.machine_id.replace(r"'", r"'\''")
+                    service_name = 'openvpn@ovs_{0}-{1}'.format(cid, nid)
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py:1517]: Now upstart is being forsaken?! HERESY (nl: ketterij)

File: ovs/lib/storagerouter.py [hunk 9]
```diff
@@ -1429,38 +1438,139 @@ def get_proxy_config(vpool_guid, storagerouter_guid):
...
+                    if service_manager.has_service(name=service_name, client=root_client):
+                        service_manager.stop_service(name=service_name, client=root_client)
+                    root_client.file_delete(filenames=['/etc/openvpn/ovs_*'])
+
+        # Configure support agent
+        if support_agent_change is True:
+            service_name = 'support-agent'
+            Configuration.set(key=support_agent_key, value=support_agent_new)
+            for root_client in root_clients.itervalues():
+                if support_agent_new is True:
```
Comment (JeffreyDevloo) [inline ovs/lib/storagerouter.py:1527]: I don't like these names, but I'll let it slide

**Review by JeffreyDevloo — CHANGES_REQUESTED** [review]

File: ovs/extensions/support/agent.py (hunk 4 above)
Comment (kvanhijf) [reply ovs/extensions/support/agent.py:112]: Yeah, serves 'em right!!

File: ovs/extensions/support/agent.py (hunk 5 above)
Comment (kvanhijf) [reply ovs/extensions/support/agent.py]: Created ticket for this as this requires a lot more changes and testing: https://github.com/openvstorage/framework/issues/1846

File: ovs/lib/storagerouter.py (hunk 8 above)
Comment (kvanhijf) [reply ovs/lib/storagerouter.py:1517]: Same related ticket: https://github.com/openvstorage/framework/issues/1846

File: ovs/lib/storagerouter.py (hunk 9 above)
Comment (kvanhijf) [reply ovs/lib/storagerouter.py:1527]: Let it be, oh let it be
Whisper wor-ords of wisdom
...

**Review by JeffreyDevloo — APPROVED** [review]

**Pull request merged**